cache-prefetch.go

```diff
@@ -0,0 +1,183 @@
+	c := &CachePrefetch{
+		CachePrefetchOptions: opt,
+		id:                   id,
+		resolver:             resolver,
+		metrics: &CachePrefetchMetrics{
+			domainEntries: map[string]CachePrefetchEntry{},
+		},
+	}
+	if c.CacheTTLPollingCheckInterval == 0 {
+		c.CacheTTLPollingCheckInterval = time.Minute
+	}
+	if c.RecordQueryHitsMin == 0 || c.RecordQueryHitsMin == 1 || c.RecordQueryHitsMin == -1 {
+		// Set to hit after one record hit
+		// fetch opportunistically
+		c.RecordQueryHitsMin = 1
+	}
+	return c
+}
+
+func (r *CachePrefetch) Resolve(q *dns.Msg, ci ClientInfo) (*dns.Msg, error) {
+	if len(q.Question) < 1 {
+		return nil, errors.New("no question in query")
+	}
+	// While multiple questions in one DNS message is part of the standard,
+	// it's not actually supported by servers. If we do get one of those,
+	// just pass it through and bypass caching.
+	if len(q.Question) > 1 {
+		return r.resolver.Resolve(q, ci)
+	}
+
+	go r.requestAddPrefetchJob(q, ci)
+
+	// Get a response from upstream
+	a, err := r.resolver.Resolve(q.Copy(), ci)
+	if err != nil || a == nil {
+		return nil, err
+	}
+
+	// Put the upstream response into the cache and return it. Need to store
+	// a copy since other elements might modify the response, like the replacer.
+	return a, nil
+}
+
+func (r *CachePrefetch) String() string {
+	return r.id
+}
+func (r *CachePrefetch) startCachePrefetch(q *dns.Msg, ci ClientInfo) {
+	var maxNumberOfErrorsBeforeDiscardingPrefetchJob = int64(5)
+	var qname = q.Question[0].Name
+
+	for {
+		var domainEntry = r.metrics.domainEntries[qname]
+		if domainEntry.prefetchingStatus.Value() == 1 { // only prefetch if status is 1
+			time.Sleep(r.CacheTTLPollingCheckInterval)
+			a, err := r.resolver.Resolve(q.Copy(), ci)
+			if err != nil || a == nil {
+				Log.WithFields(logrus.Fields{"err": err}).Trace("prefetch error")
+				domainEntry.errorCount.Add(1)
+				r.metrics.domainEntries[qname] = domainEntry
+			} else if domainEntry.errorCount.Value() > 0 {
+				// reset error count after a successful request
+				domainEntry.errorCount.Set(0)
+				r.metrics.domainEntries[qname] = domainEntry
+			}
+
+			if domainEntry.errorCount.Value() >= maxNumberOfErrorsBeforeDiscardingPrefetchJob {
+				// We don't want a bunch of error based prefetch jobs so after a certain number of errors we discard request
+				// TODO discard error prone jobs @frank? How do I do that
+				Log.WithFields(logrus.Fields{"errorCount": domainEntry.errorCount.Value(), "qname": qname}).Trace("prefetch disabled")
+				domainEntry.prefetchingStatus.Set(2)
+				r.metrics.domainEntries[qname] = domainEntry
+			}
+		} else if  domainEntry.prefetchingStatus.Value() == 2 {
+			break
+		}
+	}
+}
+func (r *CachePrefetch) requestAddPrefetchJob(q *dns.Msg, ci ClientInfo) {
+	if len(q.Question) < 1 {
+		return
+	}
+	var qname = q.Question[0].Name
+	r.metrics.domainEntries[qname] = CachePrefetchEntry{}
+	var domainEntry = r.metrics.domainEntries[qname]
```

Comment:
Not sure I understand what this is trying to do, first it flushes the record with a blank one, then it reads that empty struct into `domainEntry`?

Reply:
I was trying to initiatize it